README.md
<!---
SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
SPDX-License-Identifier: MIT
-->

# rusty-man

```
$ rusty-man rusty-man | head -5
rusty-man                       Module rusty-man                       rusty-man

DESCRIPTION
      rusty-man is a command-line viewer for documentation generated by
      `rustdoc`.
```

<img alt="A screenshot of rusty-man displaying the std::result documentation with the tui viewer"
     src="https://git.sr.ht/~ireas/rusty-man/blob/master/rusty-man.jpg"/>

## Example Usage

Before running rusty-man, make sure that you generated the documentation with
`cargo doc`:
```
$ cd rusty-man && cargo doc
```

Then you can open the documentation for an item:
```
$ rusty-man kuchiki::NodeRef
```

You don’t have to specificy the full item name:
```
$ rusty-man NodeRef
```

If the documentation is not placed in `./target/doc` or the Rust installation
directory, you have to manually set the source path using the `-s`/`--source`
option:
```
$ rusty-man --source my/other/crate/target/doc io::Error
```

If there are multiple matches for the keyword, rusty-man will show you a list
of all matching items:
```
$ rusty-man --source my/other/crate/target/doc u8
Found mulitple matches for u8 – select one of:

[ 0 ] core::u8: The 8-bit unsigned integer type.
[ 1 ] std::u8: The 8-bit unsigned integer type.

> 1
```

For more usage information, run `rusty-man --help`.

## Viewers

rusty-man has three viewers (select one with the `--viewer` option) that
determine the output format:
- `plain` prints the documentation to the standard output without formatting.
  This is the default for non-interactive use.
- `rich` prints the documentation to the standard output with formatting and
  optional syntax highlighting.  This is the default for interactive use.
- `tui` is an interactive terminal interface.  You can read the documentation
  and follow links.

## Installation

To install rusty-man from source, clone this repository and use cargo to build
the project, or use `cargo install`.  See the [`INSTALL.md`][] file for more
information.  rusty-man is also available as a package for these distributions:
- Arch Linux: [`rusty-man`][pkg-aur] in the Arch User Repository

rusty-man is developed for Unix-like systems.  It should run on other platforms
too, but with some limitations:  The `rich` and `tui` viewers uses ANSI escape
codes, which are not supported on older Windows versions.  If you have trouble
running rusty-man on your platform or if you want to help porting rusty-man to
other platforms, please let me know.

## Integrations

### vim

You can set rusty-man as the keyword program with `set keywordprg=rusty-man`
and then use `K` to open the documentation for the item under the cursor.  If
you want to apply this setting for all Rust files, add it to the
`ftplugin/rust.vim` file in your vim runtime directory (e. g. `~/.vim`).

## Contributing

Contributions to this project are very welcome!  You can contribute by writing
code or documentation or by testing.  See the [`CONTRIBUTING.md`][] file for
more information.

If you are looking for a good starting point, have a look at the [issues with
the label “good first issue”][issues] in rusty-man’s issue tracker.

## Minimum Supported Rust Version

This crate supports Rust 1.45.0 or later.

## Contact

For bug reports, patches, feature requests or other messages, please send a
mail to the mailing list [~ireas/rusty-man-dev@lists.sr.ht][] ([public
archive][]) or [contact me directly][].  You can also use the rusty-man [issue
tracker][] for bug reports or features requests.

## License

This project is licensed under the [MIT License][].  For license information
for the data used by the test suite, see the [`.reuse/dep5`][] file.

`rusty-man` complies with [version 3.0 of the REUSE specification][reuse].

[`crossterm`]: https://lib.rs/crossterm
[`INSTALL.md`]: https://git.sr.ht/~ireas/rusty-man/tree/refs/heads/master/INSTALL.md
[`CONTRIBUTING.md`]: https://git.sr.ht/~ireas/rusty-man/tree/refs/heads/master/CONTRIBUTING.md
[issues]: https://todo.sr.ht/~ireas/rusty-man?search=label:%22good%20first%20issue%22%20status%3Aopen
[~ireas/rusty-man-dev@lists.sr.ht]: mailto:~ireas/rusty-man-dev@lists.sr.ht
[contact me directly]: mailto:robin.krahl@ireas.org
[public archive]: https://lists.sr.ht/~ireas/rusty-man-dev
[issue tracker]: https://todo.sr.ht/~ireas/rusty-man
[MIT license]: https://opensource.org/licenses/MIT
[`.reuse/dep5`]: https://git.sr.ht/~ireas/rusty-man/tree/refs/heads/master/.reuse/dep5
[reuse]: https://reuse.software/practices/3.0/

[pkg-aur]: https://aur.archlinux.org/packages/rusty-man/


CHANGELOG.md
<!---
SPDX-FileCopyrightText: 2020-2021 Robin Krahl <robin.krahl@ireas.org>
SPDX-License-Identifier: MIT
-->

# Changelog for rusty-man

## Unreleased

- Add --open option
- Fix html parse method to support latest rustdoc
- Add support for toolchain >= 1.69

## v0.5.0 (2021-10-26)

This minor release adds support for Rust 1.54.0, 1.55.0 and 1.56.0 and bumps
the MSRV to 1.45.0.

- Scroll by page on PageUp/PageDown in tui viewer.
- Bump MSRV to 1.45.0.
- Add support and tests for Rust 1.54.0, 1.55.0 and 1.56.0.

## v0.4.3 (2021-06-19)

This patch release adds support for Rust 1.53.0 and for non-default target
directories.

- Prepare supporting multiple parser backends.
- Read the `$CARGO_TARGET_DIR` and `$CARGO_BUILD_TARGET_DIR` environment
  variables to determine the Cargo target directory.
- Add support and tests for Rust 1.53.0.

## v0.4.2 (2021-06-06)

This patch releases adds support for the new search index format (Rust 1.52.0
and later).

- Add `o` command for opening a documentation item to the tui viewer.
- Add tests for Rust 1.48.0, 1.49.0, 1.50.0, 1.51.0, 1.52.0 and 1.52.1.
- Add support for the new search index format introduced in Rust 1.52.0.

## v0.4.1 (2020-10-11)

This patch release fixes an issue with the pager configuration.

- Use the `LESS` environment variable to set the options for the pager.
- Add the `--pager` option to set the pager for the plain and rich viewers.

## v0.4.0 (2020-10-09)

This minor release introduces a new interactive viewer, tui.  It also adds
syntax highlighting for code in the documentation and support for Rust 1.47.0.

- Remove suffix from duplicate members (e. g. from Deref implementations).
- Use the `merge` crate to merge the command-line arguments and the settings in
  the configuration file.
  - Add `merge` dependency in version 0.1.0.
- Add syntax highlighting for code snippets in the doc comments.
- Add interactive tui viewer that adds support for following links from the
  documentation.
- Add tests for Rust 1.47.0.
- Remove Notable Traits section from definitions.

## v0.3.0 (2020-09-11)

This minor release adds support for Rust 1.46.0 and significantly improves the
test suite.

- Improve handling of different items with same name:
  - Add the item type to the item list if multiple matches are found in the
    search index.
  - Respect the item type when opening the documentation for an item that has
    been found in the search index.
- Refactor test suite:
  - Store the documentation generated by all supported rustdoc versions in the
    `tests/html` directory.
  - Use one snapshot per test case and rustdoc version.
- Refactor member lookup for compatibility with Rust 1.46.0.
- Add tests for Rust 1.46.0.
- Improve test suite:
  - Add test for `Parser::find_member`.
  - Add test for `Index::find`.

## v0.2.0 (2020-08-11)

This minor release adds support for syntax highlighting of code snippets and
for configuration files.

- Add syntax highlighting:
  - Add syntax highlighting using `syntect` for code snippets displayed with
    the rich text viewer.
  - Add the `--no-syntax-highlight` option to disable syntax highlighting.
  - Add the `--theme [theme]` option to select the syntax highlighting theme.
- Add support for configuration files:
  - Load the `config.toml` file from the config directory according to the XDG
    Base Directory Specification `${XDG_CONFIG_HOME}/rusty-man/config.toml`,
    where `${XDG_CONFIG_HOME}` defaults to `${HOME}/.config`.  The
    configuration file can be used to set defaults for the command-line
    options.
  - Add the `--config-file [file]` option to set a custom configuration file.
- Add the `--width [width]` option to set a fixed output width and the
  `--max-width [max]` option to set the maximum output width.
- Improve line break rendering when displaying code.
- Add integration test suite.

## v0.1.3 (2020-07-28)

This patch release adds support for documentation generated with Rust 1.45.0
and fixes some minor bugs in the documentation parser.  It also adds the
documentation downloaded using rustup to the default sources.

- Use `rustc --print sysroot` to determine the Rust installation directory
  instead of always using `/usr`.
- Improve the documentation parser:
  - Fix the definition of methods to only contain the actual definition.
  - Remove spurious members in module documentation.
  - Show the definition for constants and typedefs.
  - Fix group and ID for typdef items.
  - Extract the description of module items as HTML instead of plain text.
  - Sort implementations alphabetically.
  - Fix list of methods and trait implementations for Rust 1.45.

## v0.1.2 (2020-07-25)

This patch release adds basic logging output and a new `-e`/`--examples` option
to extract only the examples from the documentation.  It also fixes a bug when
displaying the documentation for a function.

- Add basic logging using `env_logger` that can be enabled by setting the
  environment variable `RUST_LOG=info`.
  - Add `env_logger` dependency in version 0.7.1.
  - Add `log` dependency in version 0.4.11.
- Show the definition for global functions.
- Add the `-e`/`--examples` option to only show the examples instead of opening
  the full documentation for an item.

## v0.1.1 (2020-07-24)

This patch release fixes some minor issues with the documentation displayed on
crates.io.

## v0.1.0 (2020-07-24)

Initial release with support for directory sources and including viewers for
plain and rich text.


src/doc.rs
// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: MIT

use std::collections;
use std::convert;
use std::fmt;
use std::ops;
use std::str;

use crate::parser::html;

#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct Name {
    // s[..first_end] == first
    // s[last_name_start..] == last_name
    s: String,
    first_end: usize,
    last_start: usize,
}

#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct Fqn(Name);

#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct Text {
    pub plain: String,
    pub html: String,
}

#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct Code(String);

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum ItemType {
    // module members
    ExternCrate,
    Import,
    Primitive,
    Module,
    Macro,
    Struct,
    Enum,
    Constant,
    Static,
    Trait,
    Function,
    Typedef,
    Union,

    // struct and union members
    StructField,

    // enum members
    Variant,

    // associated items
    AssocType,
    AssocConst,
    Method,
    Impl,

    // other items
    TyMethod,
    ForeignType,
    Keyword,
    OpaqueTy,
    ProcAttribute,
    ProcDerive,
    TraitAlias,
}

#[derive(Copy, Clone, Debug)]
pub struct ConvertError;

impl core::fmt::Display for ConvertError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("Convert error")
    }
}

impl std::error::Error for ConvertError {}

impl core::convert::TryFrom<char> for ItemType {
    type Error = ConvertError;

    fn try_from(value: char) -> Result<Self, Self::Error> {
        use core::convert::TryInto;
        Ok((value as u8 - 'A' as u8).try_into()?)
    }
}

impl core::convert::TryFrom<u8> for ItemType {
    type Error = ConvertError;
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(ItemType::Module),
            1 => Ok(ItemType::ExternCrate),
            2 => Ok(ItemType::Import),
            3 => Ok(ItemType::Struct),
            4 => Ok(ItemType::Enum),
            5 => Ok(ItemType::Function),
            6 => Ok(ItemType::Typedef),
            7 => Ok(ItemType::Static),
            8 => Ok(ItemType::Trait),
            9 => Ok(ItemType::Impl),
            10 => Ok(ItemType::TyMethod),
            11 => Ok(ItemType::Method),
            12 => Ok(ItemType::StructField),
            13 => Ok(ItemType::Variant),
            14 => Ok(ItemType::Macro),
            15 => Ok(ItemType::Primitive),
            16 => Ok(ItemType::AssocType),
            17 => Ok(ItemType::Constant),
            18 => Ok(ItemType::AssocConst),
            19 => Ok(ItemType::Union),
            20 => Ok(ItemType::ForeignType),
            21 => Ok(ItemType::Keyword),
            22 => Ok(ItemType::OpaqueTy),
            23 => Ok(ItemType::ProcAttribute),
            24 => Ok(ItemType::ProcDerive),
            25 => Ok(ItemType::TraitAlias),
            _ => Err(ConvertError),
        }
    }
}

#[derive(Clone, Debug)]
pub struct Doc {
    pub name: Fqn,
    pub ty: ItemType,
    pub description: Option<Text>,
    pub definition: Option<Code>,
    pub groups: collections::BTreeMap<ItemType, Vec<MemberGroup>>,
    pub url: Option<String>,
}

#[derive(Clone, Debug)]
pub struct MemberGroup {
    pub title: Option<String>,
    pub members: Vec<Doc>,
}

#[derive(Clone, Debug)]
pub struct Example {
    pub description: Option<Text>,
    pub code: Code,
}

impl Name {
    pub fn is_singleton(&self) -> bool {
        self.last_start == 0
    }

    pub fn first(&self) -> &str {
        &self.s[..self.first_end]
    }

    pub fn last(&self) -> &str {
        &self.s[self.last_start..]
    }

    pub fn full(&self) -> &str {
        &self.s
    }

    pub fn rest(&self) -> Option<&str> {
        if self.is_singleton() {
            None
        } else {
            Some(&self.s[self.first_end + 2..])
        }
    }

    pub fn rest_or_first(&self) -> &str {
        self.rest().unwrap_or_else(|| self.first())
    }

    pub fn parent(&self) -> Option<Self> {
        if self.is_singleton() {
            None
        } else {
            Some((&self.s[..self.last_start - 2]).to_owned().into())
        }
    }

    pub fn child(&self, s: &str) -> Self {
        let mut name = self.s.clone();
        name.push_str("::");
        name.push_str(s);
        name.into()
    }

    pub fn ends_with(&self, name: &Name) -> bool {
        self.s == name.s || self.s.ends_with(&format!("::{}", name.s))
    }
}

impl AsRef<str> for Name {
    fn as_ref(&self) -> &str {
        self.s.as_ref()
    }
}

impl From<String> for Name {
    fn from(s: String) -> Self {
        let first_end = s.find("::").unwrap_or_else(|| s.len());
        let last_start = s.rfind("::").map(|i| i + 2).unwrap_or(0);
        Self {
            s,
            first_end,
            last_start,
        }
    }
}

impl From<Name> for String {
    fn from(n: Name) -> Self {
        n.s
    }
}

impl fmt::Display for Name {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.s)
    }
}

impl str::FromStr for Name {
    type Err = convert::Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(s.to_owned().into())
    }
}

impl Fqn {
    pub fn krate(&self) -> &str {
        self.first()
    }

    pub fn parent(&self) -> Option<Self> {
        self.0.parent().map(From::from)
    }

    pub fn child(&self, s: &str) -> Self {
        self.0.child(s).into()
    }
}

impl AsRef<str> for Fqn {
    fn as_ref(&self) -> &str {
        self.0.as_ref()
    }
}

impl From<Name> for Fqn {
    fn from(n: Name) -> Self {
        Self(n)
    }
}

impl From<String> for Fqn {
    fn from(s: String) -> Self {
        Self(s.into())
    }
}

impl fmt::Display for Fqn {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.0)
    }
}

impl ops::Deref for Fqn {
    type Target = Name;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Code {
    pub fn new(s: String) -> Code {
        Code(s)
    }
}

impl fmt::Display for Code {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.0)
    }
}

impl ops::Deref for Code {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl ItemType {
    pub fn name(&self) -> &str {
        match self {
            ItemType::Module => "Module",
            ItemType::ExternCrate => "Extern Crate",
            ItemType::Import => "Import",
            ItemType::Struct => "Struct",
            ItemType::Enum => "Enum",
            ItemType::Function => "Function",
            ItemType::Typedef => "Typedef",
            ItemType::Static => "Static",
            ItemType::Trait => "Trait",
            ItemType::Impl => "Implementation",
            ItemType::TyMethod => "Required Method",
            ItemType::Method => "Method",
            ItemType::StructField => "Field",
            ItemType::Variant => "Variant",
            ItemType::Macro => "Macro",
            ItemType::Primitive => "Primitive",
            ItemType::AssocType => "Associated Type",
            ItemType::Constant => "Constant",
            ItemType::AssocConst => "Associated Const",
            ItemType::Union => "Union",
            ItemType::ForeignType => "Foreign Type",
            ItemType::Keyword => "Keyword",
            ItemType::OpaqueTy => "Opaque Type",
            ItemType::ProcAttribute => "Proc Attribute",
            ItemType::ProcDerive => "Proc Derive",
            ItemType::TraitAlias => "Trait Alias",
        }
    }

    pub fn group_name(&self) -> &str {
        match self {
            ItemType::Module => "Modules",
            ItemType::ExternCrate => "Extern Crates",
            ItemType::Import => "Imports",
            ItemType::Struct => "Structs",
            ItemType::Enum => "Enums",
            ItemType::Function => "Functions",
            ItemType::Typedef => "Typedefs",
            ItemType::Static => "Statics",
            ItemType::Trait => "Traits",
            ItemType::Impl => "Implementations",
            ItemType::TyMethod => "Required Methods",
            ItemType::Method => "Methods",
            ItemType::StructField => "Fields",
            ItemType::Variant => "Variants",
            ItemType::Macro => "Macros",
            ItemType::Primitive => "Primitives",
            ItemType::AssocType => "Associated Types",
            ItemType::Constant => "Constants",
            ItemType::AssocConst => "Associated Consts",
            ItemType::Union => "Unions",
            ItemType::ForeignType => "Foreign Types",
            ItemType::Keyword => "Keywords",
            ItemType::OpaqueTy => "Opaque Types",
            ItemType::ProcAttribute => "Proc Attributes",
            ItemType::ProcDerive => "Proc Derives",
            ItemType::TraitAlias => "Trait Aliases",
        }
    }
}

impl str::FromStr for ItemType {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "mod" => Ok(ItemType::Module),
            "externcrate" => Ok(ItemType::ExternCrate),
            "import" => Ok(ItemType::Import),
            "struct" => Ok(ItemType::Struct),
            "union" => Ok(ItemType::Union),
            "enum" => Ok(ItemType::Enum),
            "fn" => Ok(ItemType::Function),
            "type" => Ok(ItemType::Typedef),
            "static" => Ok(ItemType::Static),
            "trait" => Ok(ItemType::Trait),
            "impl" => Ok(ItemType::Impl),
            "tymethod" => Ok(ItemType::TyMethod),
            "method" => Ok(ItemType::Method),
            "structfield" => Ok(ItemType::StructField),
            "variant" => Ok(ItemType::Variant),
            "macro" => Ok(ItemType::Macro),
            "primitive" => Ok(ItemType::Primitive),
            "associatedtype" => Ok(ItemType::AssocType),
            "constant" => Ok(ItemType::Constant),
            "associatedconstant" => Ok(ItemType::AssocConst),
            "foreigntype" => Ok(ItemType::ForeignType),
            "keyword" => Ok(ItemType::Keyword),
            "opaque" => Ok(ItemType::OpaqueTy),
            "attr" => Ok(ItemType::ProcAttribute),
            "derive" => Ok(ItemType::ProcDerive),
            "traitalias" => Ok(ItemType::TraitAlias),
            _ => Err(anyhow::anyhow!("Unsupported item type: {}", s)),
        }
    }
}

impl Doc {
    pub fn new(name: Fqn, ty: ItemType) -> Self {
        Self {
            name,
            ty,
            description: Default::default(),
            definition: Default::default(),
            groups: Default::default(),
            url: None,
        }
    }

    pub fn find_examples(&self) -> anyhow::Result<Vec<Example>> {
        if let Some(description) = &self.description {
            html::Parser::from_string(&description.html)?.find_examples()
        } else {
            Ok(Vec::new())
        }
    }

    pub fn set_url(&mut self, path: &std::path::Path, hash: Option<String>) {
        let mut path = path
            .canonicalize()
            .unwrap()
            .as_os_str()
            .to_string_lossy()
            .to_string();
        if path.starts_with("/") {
            path = (&path[1..]).to_string();
        }

        if let Some(hash) = hash {
            self.url = Some(format!("file:///{}#{}", path, hash));
        } else {
            self.url = Some(format!("file:///{}", path));
        }
    }
}

impl fmt::Display for Doc {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if let Some(description) = &self.description {
            write!(f, "{}: {}", &self.name, description.plain)
        } else {
            write!(f, "{}", &self.name)
        }
    }
}

impl MemberGroup {
    pub fn new(title: Option<String>) -> Self {
        MemberGroup {
            title,
            members: Vec::new(),
        }
    }
}

impl Example {
    pub fn new(description: Option<Text>, code: Code) -> Self {
        Example { description, code }
    }
}

#[cfg(test)]
mod tests {
    use super::Name;

    fn assert_name(input: &str, first: &str, last: &str, rest: &str) {
        let name: Name = input.to_owned().into();
        assert_eq!(first, name.first(), "first for '{}'", input);
        assert_eq!(last, name.last(), "last for '{}'", input);
        assert_eq!(input, name.full(), "full for '{}'", input);
        if rest == input {
            assert_eq!(None, name.rest(), "rest for '{}'", input);
        } else {
            assert_eq!(Some(rest), name.rest(), "rest for '{}'", input);
        }
    }

    #[test]
    fn test_empty_name() {
        assert_name("", "", "", "");
    }

    #[test]
    fn test_crate_name() {
        assert_name("rand", "rand", "rand", "rand");
    }

    #[test]
    fn test_module_name() {
        assert_name("rand::error", "rand", "error", "error");
        assert_name("rand::error::nested", "rand", "nested", "error::nested");
    }

    #[test]
    fn test_item_name() {
        assert_name("rand::Error", "rand", "Error", "Error");
        assert_name("rand::error::Error", "rand", "Error", "error::Error");
    }

    #[test]
    fn test_member_name() {
        assert_name("rand::Error::source", "rand", "source", "Error::source");
        assert_name(
            "rand::error::Error::source",
            "rand",
            "source",
            "error::Error::source",
        );
    }

    #[test]
    fn test_colon() {
        assert_name("er:ror::Error", "er:ror", "Error", "Error");
    }
}
